File: 207-course-schedule/course-schedule.cpp
class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {

        vector<int>numDependencies(numCourses, 0); // how many dependencies for each course
        vector<vector<int>>reverseDep(numCourses, vector<int>()); 
        
        for(vector<int>& pre : prerequisites)
        {
            int x = pre[0], y = pre[1];
            numDependencies[x]++;
            reverseDep[y].push_back(x);
        }
        queue<int>q;
        for(int i = 0 ; i < numCourses ; i++)
        {
            if(numDependencies[i] == 0){
                q.push(i);
            }
        }
        int ans = 0;
        while(!q.empty())
        {
            int size_q = q.size();
            ans+=size_q;
            for(int i = 0 ; i < size_q ; i++)
            {
                int course = q.front(); q.pop();
                for(int v : reverseDep[course])
                {
                    numDependencies[v]-=1;
                    if(numDependencies[v] == 0)
                    {
                        q.push(v);
                    }
                }
            }
        }
        return ans == numCourses;
    }
};

File: 207-course-schedule/course-schedule.java
class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        int[] outdegree = new int[numCourses];
        List<List<Integer>> adjList = new ArrayList<>();
        for (int i = 0; i < numCourses; i++) {
            adjList.add(new ArrayList<>());
        }
        for (int[] courses : prerequisites) {
            int index = courses[0];
            int course = courses[1];
            adjList.get(course).add(index);
            outdegree[index]++;
        }
        Queue<Integer> q = new ArrayDeque<>();
        for (int i = 0; i < numCourses; i++) {
            if (outdegree[i] == 0) {
                q.offer(i);
            }
        }
        int count_courses = 0;
        while (!q.isEmpty()) {
            int course = q.poll();
            count_courses++;
            for (int c : adjList.get(course)) {
                outdegree[c]--;
                if (outdegree[c] == 0)
                    q.offer(c);
            }
        }
        return count_courses == numCourses;
    }
}